runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,905 @@
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Dispute slashing pallet.
+//!
+//! The implementation relies on the `offences` pallet and
+//! looks like a hybrid of `im-online` and `grandpa` equivocation handlers.
+//! Meaning, we submit an `offence` for the concluded disputes about
+//! the current session candidate directly from the runtime.
+//! If, however, the dispute is about a past session, we record pending
+//! slashes on chain, without `FullIdentification` of the offenders.
+//! Later on, a block producer can submit an unsigned transaction with
+//! `KeyOwnershipProof` of an offender and submit it to the runtime
+//! to produce an offence.
+//! The reason for this separation is that we do not want to rely on
+//! `FullIdentification` to be available on chain for the past sessions
+//!  because it is quite heavy.
+
+use crate::{
+	disputes,
+	initializer::ValidatorSetCount,
+	session_info::{AccountId, IdentificationTuple},
+};
+use frame_support::{
+	traits::{Get, KeyOwnerProofSystem, ValidatorSet, ValidatorSetWithIdentification},
+	weights::{Pays, Weight},
+};
+use parity_scale_codec::{Decode, Encode};
+use primitives::v2::{CandidateHash, SessionIndex, ValidatorId, ValidatorIndex};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::Convert,
+	transaction_validity::{
+		InvalidTransaction, TransactionPriority, TransactionSource, TransactionValidity,
+		TransactionValidityError, ValidTransaction,
+	},
+	DispatchResult, KeyTypeId, Perbill, RuntimeDebug,
+};
+use sp_session::{GetSessionNumber, GetValidatorCount};
+use sp_staking::offence::{DisableStrategy, Kind, Offence, OffenceError, ReportOffence};
+use sp_std::{
+	collections::btree_map::{BTreeMap, Entry},
+	prelude::*,
+};
+
+const LOG_TARGET: &str = "runtime::slashing";
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking;
+
+/// Timeslots should uniquely identify offences
+/// and are used for the offence deduplication.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+pub struct DisputesTimeSlot {
+	// The order of these matters for `derive(Ord)`.
+	session_index: SessionIndex,
+	candidate_hash: CandidateHash,
+}
+
+impl DisputesTimeSlot {
+	pub fn new(session_index: SessionIndex, candidate_hash: CandidateHash) -> Self {
+		Self { session_index, candidate_hash }
+	}
+}
+
+/// An offence that is filed when a series of validators lost a dispute
+/// about an invalid candidate.
+#[derive(RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct ForInvalidOffence<KeyOwnerIdentification> {
+	/// The size of the validator set in that session.
+	pub validator_set_count: ValidatorSetCount,
+	/// Should be unique per dispute.
+	pub time_slot: DisputesTimeSlot,
+	/// Staking information about the validators that lost the dispute
+	/// needed for slashing.
+	pub offenders: Vec<KeyOwnerIdentification>,
+}
```

Comment:
It was initially one struct, but then I experimented with exponential slashing and there became different. I plan to introduce exponential slashing a bit later and get a simple version reviewed and merged first.

Reply:
Are you suggesting to also have the same `const ID: Kind` for these two types of offences?

In theory it could happen that a dispute can conclude both "against valid" and "for invalid". But for that to happen we'd need 2f+1 malicious validators I guess? At this point all bets are off. A bigger worry if that can happen due to a bug.

But yes, it's possible to have only one struct with some [small changes](https://github.com/paritytech/substrate/compare/ao-make-slash-fraction-non-static) in substrate. But I'm not sure about the benefits of going this way other than code dedup.

Reply:
I've merged the offence types to simplify the code, but my concern still stays.